internal/argocd/revision.go

```diff
@@ -10,68 +11,163 @@ import (
 		}
-		if targetPromoMechanism == nil {
-			targetPromoMechanism = update
+
+		desiredRevision, err := getRevisionFromSource(ctx, cl, stage, s, desiredOrigin, frght)
+
+		if err != nil {
+			return nil, err
 		}
-		desiredOrigin := freight.GetDesiredOrigin(stage, targetPromoMechanism)
-		repoURL := app.Spec.Source.RepoURL
-		chartName := app.Spec.Source.Chart
-		if !strings.Contains(repoURL, "://") {
-			// In Argo CD ApplicationSource, if a repo URL specifies no protocol and a
-			// chart name is set (already confirmed at this point), we can assume that
-			// the repo URL is an OCI registry URL. Kargo Warehouses and Freight,
-			// however, do use oci:// at the beginning of such URLs.
-			//
-			// Additionally, where OCI is concerned, an ApplicationSource's repoURL is
-			// really a registry URL, and the chart name is a repository within that
-			// registry. Warehouses and Freight, however, handle things more correctly
-			// where a repoURL points directly to a repository and chart name is
-			// irrelevant / blank. We need to account for this when we search our
-			// Freight for the chart.
-			repoURL = fmt.Sprintf(
-				"oci://%s/%s",
-				strings.TrimSuffix(repoURL, "/"),
-				chartName,
-			)
-			chartName = ""
+
+		if desiredRevision != "" {
+			appLogger.WithValues("source", s,
+				"revision", desiredRevision).Debug("Found desired revision for application source.")
+			return []string{desiredRevision}, nil
 		}
+
+		appLogger.WithValues("source", s).Debug("Could not determine desired revision for application from this source.")
+		return nil, nil
+	}
+
+	// An application may have more than one source that points to the same Git repository,
+	// eg. a Helm and a vanilla manifest source.
+	// In that situation it doesn't matter which source's target revision is returned as ArgoCD does not support
+	// different target revisions for sources targeting the same repository.
+	var revisions = make([]string, len(app.Spec.Sources))
+	isSynced := app.Status.Sync.Revisions != nil
+
+	// ArgoCD uses a shadow-array to store the corresponding revisions preserving the order of app.spec.sources.
+	// We match that logic here and return the desired revision for each source,
+	// or empty string if one is not found.
+	for i, src := range app.Spec.Sources {
+		s := src
+
+		syncedRevisionOfSource := ""
+		if isSynced && i < len(app.Status.Sync.Revisions) {
+			syncedRevisionOfSource = app.Status.Sync.Revisions[i]
+		}
+
+		// If there is a source update that targets app.Spec.Source, it might
+		// have its own ideas about the desired revision.
+		targetUpdate := getTargetUpdate(ctx, update, &s)
+		desiredOrigin := freight.GetDesiredOrigin(ctx, stage, targetUpdate)
+		if desiredOrigin == nil {
+			appLogger.WithValues("source", s,
+				"revision", syncedRevisionOfSource).Debug("Could not determine desired origin" +
+				" for application source, using existing revision.")
+			revisions[i] = syncedRevisionOfSource
```

Comment:
This doesn't look right to me. The previous behavior of this function had been to find the (singular) desired revision for the single source App and it returned an empty string if it couldn't determine / had no opinion about the revision we should be synced to. I didn't expect that to change when this was updated to work with multiple sources/revisions. If we can't figure out what's desired for a particular source, the corresponding element in the array of desired revisions should just be an empty string.